accepted/00000-saltclass-external-pillar.md

```diff
@@ -0,0 +1,147 @@
+
+### Changes for image pillars are
+
+1) Adapt SaltStateGeneratorService.java to store pillar data as a `saltclass` `class` and enable it for all minions
+2) Migrate existing image pillars to saltclass classes
+
+## Step 3: Migrating existing pillar data to use satlclass
+
+Migrating existing pillar data will require two steps:
+1) Modification of java code storing minion pillars:
+  - drop `pillar_` prefix from pillar files
+  - store pillar data under `pillar` key
+2) Rename existing pillar files from `pillar_{minion_id}.yml` to `{minion_id}.yml`
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Big refactoring underway, but stalled
+
+There is an upstream pull request about major `saltclass` refactoring - https://github.com/saltstack/salt/pull/52407 - however this PR seems to be stalled. This PR meant to fix several issues and fulfil couple of feature requests.
+The unknown is if the PR will be ever finished or if we can invest and try to finalize it ourselves. It is possible that the work stopped only temporarily due to changes in salt releases, or that the author being external contributor chose to abandon the effort.
+
+## Performance impact
+
+Including another pillar engine can have a negative performance impact. (#TODO measure)
+
+## More complexity means more break points
+
+Generally when running more complex code, there is bigger room for bugs. SaltClass it maintained by the Salt upstream and as of now have 19 issues opened against saltclass (counting both saltclass tops and pillar modules)
+
+## Modifying core functionality
+
+Pillar system is core functionality everything else depends on. Bugs within SaltClass can affect whole product. However proposed granular introduction of SaltClass instead of ```suma_minion``` should limit impact to only specific areas, hopefully providing enough time to catch and debug issues.
+
+## Upstream breakages, abandonment
+
+There is always some risk that upstream may decide to abandon SaltClass. So far maintaining SaltClass should be cheap, due to upstream involvement, but if the upstream decides for any reason to abandon SaltClass the cost of maintenance can increase dramatically (up to impossible if upstream decides to abandon it for some deep design reasons). As a mitigation there is a possibility to use [reclass](https://reclass.pantsfullofunix.net/) upon which SaltClass is based and shares storage details (see below).
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Implement custom jinja processing of pillars
+
+The major driver for `saltclass` is the ability to preparse pillars with jinja and also to reference pillars between each other. This can be implemented within existing `suma_minion.py` external pillar as well. Using `saltclass` saves us the development time in exchange for development time on integration, but given `saltclass` is an upstream project it in theory can help with maintenance.
```

Comment:
I'd really like to see valid example for why I'd need the Jinja preparsing for pillars. As mentioned before, this could have severe performance impacts. It also makes it harder to use third party providers for pillars (e.g. AD or LDAP) when customers integrate with other backend systems, as such complex pillar structures are very hard to get right.
Personally I've always seen the main advantage of saltclass/reclass in its ability to inherit and overload defaults, thus reducing the number of pillars I have to duplicate.